In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib Inline

In [2]:
data = pd.read_csv('https://s3.amazonaws.com/thinkific/file_uploads/287654/attachments/2e3/a35/b45/Social_Network_Ads.csv')

In [3]:
data.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [4]:
data.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 15.8+ KB


In [6]:
data.drop(['User ID'] , axis = 1, inplace=True)

In [7]:
data.head(10)

Gender  Age  EstimatedSalary  Purchased
0    Male   19            19000          0
1    Male   35            20000          0
2  Female   26            43000          0
3  Female   27            57000          0
4    Male   19            76000          0
5    Male   27            58000          0
6  Female   27            84000          0
7  Female   32           150000          1
8    Male   25            33000          0
9  Female   35            65000          0

In [8]:
data = pd.get_dummies(data,drop_first=True)

In [9]:
#Just rearranged the colummns
cols = ['Age','EstimatedSalary','Gender_Male','Purchased']
data= data[cols]
data.head()

Age  EstimatedSalary  Gender_Male  Purchased
0   19            19000            1          0
1   35            20000            1          0
2   26            43000            0          0
3   27            57000            0          0
4   19            76000            1          0

# Creating model using Estimated Salary and Purchased

Predicting whether a particular person would click on a social media Ad or not based on his Age,Salary,Gender



## Defining X and Y

In [10]:
X = data.iloc[:,:-1].values
X.shape

(400, 3)

In [11]:
Y = data.iloc[:,-1].values
Y.shape

(400,)

In [12]:
data.head(5)

Age  EstimatedSalary  Gender_Male  Purchased
0   19            19000            1          0
1   35            20000            1          0
2   26            43000            0          0
3   27            57000            0          0
4   19            76000            1          0

As we can see that Age and EstimatedSalary have diffrent scales i.e Age is in 10's and EstimatedSalary is in 1000's So we will scale this data for further engineering.So that they can be in same scale.

## Creating the model without Scaling


In [13]:
#Splitting the data in train and test
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=0)

In [14]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train,Y_train)

LogisticRegression()

In [15]:
Y_pred= model.predict(X_test)

In [16]:
Y_pred

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [17]:
#PERFORMANCE METRICS
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(Y_test,Y_pred)
print(confusion)

[[58  0]
 [22  0]]


In [18]:
TN = confusion[0,0]
FP = confusion[0,1]
FN = confusion[1,0]
TP = confusion[1,1]

In [19]:
print(confusion)
print("TN: ",TN)
print("FP: ",FP)
print("FN: ",FN)
print("TP: ",TP)

[[58  0]
 [22  0]]
TN:  58
FP:  0
FN:  22
TP:  0


In [20]:
from sklearn import metrics
accuracy = metrics.accuracy_score(Y_test,Y_pred)
accuracy1 = (TN+TP)/(TN+TP+FN+FP)
print("Accuracy from metrics: ",accuracy)
print("Accuracy calculated: ",accuracy1)

Accuracy from metrics:  0.725
Accuracy calculated:  0.725


### Hence Above Accuracy i.e. 72.5 % is calculated without scaling

In [21]:
print("RECALL: ",metrics.recall_score(Y_test,Y_pred))
print("Calcualted Recall :",(TP)/(TP+FN))

RECALL:  0.0
Calcualted Recall : 0.0


In [22]:
#Speificity measures the peoportion of actual negatives that are correctly identified as such: TN/N
print("Specificity/True Negative Rate: ",(TN)/(TN+FP))

Specificity/True Negative Rate:  1.0


In [23]:
print("PRECISION: ",metrics.precision_score(Y_test,Y_pred),2)
print("PRECISION CALCUALTED: ",round(TP/float(TP+FP),2))

PRECISION:  0.0 2
PRECISION CALCUALTED:  nan


C:\Users\sarth\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
<ipython-input-23-bd7faf1963e1>:2: RuntimeWarning: invalid value encountered in true_divide
  print("PRECISION CALCUALTED: ",round(TP/float(TP+FP),2))


In [24]:
from sklearn.metrics import accuracy_score,f1_score,precision_score,recall_score,roc_auc_score
accuracy = accuracy_score(Y_test , Y_pred)
recall = recall_score(Y_test , Y_pred)
precision = precision_score(Y_test , Y_pred)
f1 = f1_score(Y_test , Y_pred)
roc_auc = roc_auc_score(Y_test , Y_pred)

print("Accuracy is:",round(accuracy,2)*100)
print("F1 Score is:",round(f1,2)*100)
print("Precision is:",round(precision,2)*100)
print("Recall is:",round(recall,2)*100)
print("Roc_Auc is:",round(roc_auc,2)*100)

Accuracy is: 72.0
F1 Score is: 0.0
Precision is: 0.0
Recall is: 0.0
Roc_Auc is: 50.0


In [25]:
#CLASSIFICATION REPORT

from sklearn.metrics import classification_report
print(classification_report(Y_test,Y_pred))

              precision    recall  f1-score   support

           0       0.72      1.00      0.84        58
           1       0.00      0.00      0.00        22

    accuracy                           0.73        80
   macro avg       0.36      0.50      0.42        80
weighted avg       0.53      0.72      0.61        80



C:\Users\sarth\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [26]:
summary = pd.DataFrame([accuracy,f1,precision,recall,roc_auc],index=['Accuracy','F1','Precision','Recall','AUC'],columns=['Without Scaling'])
summary

Without Scaling
Accuracy             0.725
F1                   0.000
Precision            0.000
Recall               0.000
AUC                  0.500

# Model After Scaling

In [27]:
X = data.iloc[:,:-1].values
X.shape

(400, 3)

In [28]:
Y = data.iloc[:,-1].values
Y.shape

(400,)

In [29]:
#Splitting the data in train and test
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=0)

In [30]:
#This is how we scale

from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train =sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

In [31]:
X_train #Now this is in exponential form

array([[ 1.92295008e+00,  2.14601566e+00,  1.02532046e+00],
       [ 2.02016082e+00,  3.78719297e-01, -9.75304830e-01],
       [-1.38221530e+00, -4.32498705e-01, -9.75304830e-01],
       [-1.18779381e+00, -1.01194013e+00, -9.75304830e-01],
       [ 1.92295008e+00, -9.25023920e-01, -9.75304830e-01],
       [ 3.67578135e-01,  2.91803083e-01, -9.75304830e-01],
       [ 1.73156642e-01,  1.46942725e-01, -9.75304830e-01],
       [ 2.02016082e+00,  1.74040666e+00,  1.02532046e+00],
       [ 7.56421121e-01, -8.38107706e-01, -9.75304830e-01],
       [ 2.70367388e-01, -2.87638347e-01, -9.75304830e-01],
       [ 3.67578135e-01, -1.71750061e-01,  1.02532046e+00],
       [-1.18475597e-01,  2.20395980e+00, -9.75304830e-01],
       [-1.47942605e+00, -6.35303205e-01, -9.75304830e-01],
       [-1.28500455e+00, -1.06988428e+00,  1.02532046e+00],
       [-1.38221530e+00,  4.07691369e-01,  1.02532046e+00],
       [-1.09058306e+00,  7.55356227e-01, -9.75304830e-01],
       [-1.47942605e+00, -2.00722133e-01

In [32]:
X_test #this is also in exponential

array([[-7.98950822e-01,  4.94607583e-01,  1.02532046e+00],
       [-2.12648508e-02, -5.77359062e-01, -9.75304830e-01],
       [-3.12897090e-01,  1.46942725e-01,  1.02532046e+00],
       [-7.98950822e-01,  2.62831011e-01, -9.75304830e-01],
       [-3.12897090e-01, -5.77359062e-01, -9.75304830e-01],
       [-1.09058306e+00, -1.44652121e+00,  1.02532046e+00],
       [-7.01740076e-01, -1.59138156e+00, -9.75304830e-01],
       [-2.15686344e-01,  2.14601566e+00,  1.02532046e+00],
       [-1.96547978e+00, -5.58617754e-02, -9.75304830e-01],
       [ 8.53631867e-01, -7.80163563e-01,  1.02532046e+00],
       [-7.98950822e-01, -6.06331134e-01,  1.02532046e+00],
       [-9.93372315e-01, -4.32498705e-01, -9.75304830e-01],
       [-1.18475597e-01, -4.32498705e-01,  1.02532046e+00],
       [ 7.59458956e-02,  2.04886868e-01,  1.02532046e+00],
       [-1.77105829e+00,  4.65635512e-01,  1.02532046e+00],
       [-6.04529329e-01,  1.36376973e+00, -9.75304830e-01],
       [-1.18475597e-01,  2.04886868e-01

In [33]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train,Y_train)

LogisticRegression()

In [34]:
Y_pred= model.predict(X_test)

In [35]:
Y_pred

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1], dtype=int64)

In [36]:
#PERFORMANCE METRICS
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(Y_test,Y_pred)
print(confusion)

[[56  2]
 [ 5 17]]


In [37]:
TN = confusion[0,0]
FP = confusion[0,1]
FN = confusion[1,0]
TP = confusion[1,1]

In [38]:
print(confusion)
print("TN: ",TN)
print("FP: ",FP)
print("FN: ",FN)
print("TP: ",TP)

[[56  2]
 [ 5 17]]
TN:  56
FP:  2
FN:  5
TP:  17


In [39]:
from sklearn import metrics
accuracy = metrics.accuracy_score(Y_test,Y_pred)
accuracy1 = (TN+TP)/(TN+TP+FN+FP)
print("Accuracy from metrics: ",accuracy)
print("Accuracy calculated: ",accuracy1)

Accuracy from metrics:  0.9125
Accuracy calculated:  0.9125


### Now after scaling the accuracy increses from 72.5% to 91.25 %

In [40]:
print("RECALL: ",metrics.recall_score(Y_test,Y_pred))
print("Calcualted Recall :",(TP)/(TP+FN))

RECALL:  0.7727272727272727
Calcualted Recall : 0.7727272727272727


In [41]:
#Speificity measures the peoportion of actual negatives that are correctly identified as such: TN/N
print("Specificity/True Negative Rate: ",(TN)/(TN+FP))

Specificity/True Negative Rate:  0.9655172413793104


In [42]:
print("PRECISION: ",metrics.precision_score(Y_test,Y_pred),2)
print("PRECISION CALCUALTED: ",round(TP/float(TP+FP),2))

PRECISION:  0.8947368421052632 2
PRECISION CALCUALTED:  0.89


In [43]:
from sklearn.metrics import accuracy_score,f1_score,precision_score,recall_score,roc_auc_score
accuracy = accuracy_score(Y_test , Y_pred)
recall = recall_score(Y_test , Y_pred)
precision = precision_score(Y_test , Y_pred)
f1 = f1_score(Y_test , Y_pred)
roc_auc = roc_auc_score(Y_test , Y_pred)

print("Accuracy is:",round(accuracy,2)*100)
print("F1 Score is:",round(f1,2)*100)
print("Precision is:",round(precision,2)*100)
print("Recall is:",round(recall,2)*100)
print("Roc_Auc is:",round(roc_auc,2)*100)

Accuracy is: 91.0
F1 Score is: 83.0
Precision is: 89.0
Recall is: 77.0
Roc_Auc is: 87.0


In [44]:
#CLASSIFICATION REPORT

from sklearn.metrics import classification_report
print(classification_report(Y_test,Y_pred))

              precision    recall  f1-score   support

           0       0.92      0.97      0.94        58
           1       0.89      0.77      0.83        22

    accuracy                           0.91        80
   macro avg       0.91      0.87      0.89        80
weighted avg       0.91      0.91      0.91        80



In [45]:
summary1 = pd.DataFrame([accuracy,f1,precision,recall,roc_auc],index=['Accuracy','F1','Precision','Recall','AUC'],columns=['With Scaling'])
summary1

With Scaling
Accuracy       0.912500
F1             0.829268
Precision      0.894737
Recall         0.772727
AUC            0.869122

In [46]:
summary_combined = pd.concat([summary,summary1],axis=1)
summary_combined

Without Scaling  With Scaling
Accuracy             0.725      0.912500
F1                   0.000      0.829268
Precision            0.000      0.894737
Recall               0.000      0.772727
AUC                  0.500      0.869122